ErgoState.execTransactions perf improvements and tests

pragmaxim opened the pull request:
Solves https://github.com/ergoplatform/ergo/issues/1524 

Getting rid of too many immutable collections copying. Also I replaced Cats `sequence` with tail recursive function.

aslesarenko commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,37 +76,50 @@ object ErgoState extends ScorexLogging {
           }
 
-        execTx(tail, vs)
+        execTx(txs.tail, vs)
```
The `tail` is only fast when txs is a List, which is not good idea in a first place.

aslesarenko commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,37 +76,50 @@ object ErgoState extends ScorexLogging {
 
-        lazy val boxes: Try[(List[ErgoBox], List[ErgoBox])] = dataBoxesTry.flatMap(db => boxesToSpendTry.map(bs => (db, bs)))
+    @tailrec
+    def execTx(allTxs: Seq[ErgoTransaction], accCostTry: ValidationResult[Long]): ValidationResult[Long] = (allTxs, accCostTry) match {
```
Tuple allocation `(allTxs, accCostTry)` for every allTxs

aslesarenko commented on src/test/scala/org/ergoplatform/nodeView/state/ErgoStateSpecification.scala:
```diff
@@ -141,4 +143,36 @@ class ErgoStateSpecification extends ErgoPropertyTest {
+    )
+
+    // no transactions are valid
+    assert(ErgoState.execTransactions(Seq.empty, stateContext)(id => Try(boxes(ByteArrayWrapper(id)))).isValid)
```
Please add tests and test vectors for cost computation.

pragmaxim replied:
Ok, I wanted to but the costs are all randomly generated in the helper methods, I have to figure out how to do it so it is not mess

aslesarenko reviewed (changes requested):
I suggest to first add tests and test vectors for costs in a separate PR.
Then make optimizations on top of that tests and make sure they don't break them.

pragmaxim commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,37 +76,50 @@ object ErgoState extends ScorexLogging {
-        lazy val dataBoxesTry: Try[List[ErgoBox]] = tx.dataInputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
+    def collectBoxesById(
+                 remainingBoxIds: Iterator[ErgoBox.BoxId],
+                 resultingBoxes: Try[VectorBuilder[ErgoBox]] = Success(new VectorBuilder[ErgoBox])
+               ): Try[IndexedSeq[ErgoBox]] = {
+      if (!remainingBoxIds.hasNext) {
+        resultingBoxes.map(_.result())
+      } else {
+        checkBoxExistence(remainingBoxIds.next()) match {
+          case Success(box) =>
+            collectBoxesById(remainingBoxIds, resultingBoxes.map(_ += box))
+          case Failure(ex) =>
+            Failure(ex)
+        }
+      }
+    }
```
It does not loop through after `Failure` as the recursive function is not called in case of first `Failure`

pragmaxim replied:
`Try` allocation is necessary I believe as obtaining the `Box` from UtxoState can simply fail or it might not be there ...

pragmaxim commented:
I addressed all concerns except the `TailRec` for collecting boxes, `Try` is kind of inevitable here, the collection is `Vector`, so while loop was very ugly and it does not have any added value here.

aslesarenko commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
+      } else {
+        checkBoxExistence(remainingBoxIds.next()) match {
+          case Success(box) =>
+            collectBoxesById(remainingBoxIds, resultingBoxes.map(_ += box))
```
ArrayBuilder will be much faster than VectorBuilder, especially in this use case, where you can use arrBuilder.sizeHint(tx.inputs.length) to do single allocations. Then appending with += will be super fast.
Just compare what happens on += in VectorBuilder and ArrayBuilder.

aslesarenko commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
+      cfor(0)(_ < transactions.length && costResult.isValid, _ + 1) { i =>
+        val validCostResult = costResult.asInstanceOf[Valid[Long]]
+        val tx = transactions(i)
+        val boxesToSpendTry: Try[IndexedSeq[ErgoBox]] = collectBoxesById(tx.inputs.iterator.map(_.boxId))
```
Here you can take advantage of knowing val n = tx.inputs.length and save O(log(n)) allocations in builder by using sizeHint method.

pragmaxim replied:
yes but `tx.inputs` and tx.dataInputs` have different types so `.map(_.boxId)` here was used for readability ...

pragmaxim replied:
Maybe we should introduce new interface for `Input` and `DataInput` with `dev boxId: BoxId` ? We could then with any input that has `boxId` ... WYT ?

aslesarenko replied:
collectBoxesById implicitly creates an empty builder here.  However, since we know the size, we can use sizeHint of the builder, to pre-allocate space.

pragmaxim replied:
but `ArrayBuilder` does not take size hints, one can either `new mutable.ArrayBuilder.ofRef[ErgoBox]()` or `mutable.ArrayBuilder.make()`

pragmaxim replied:
I see, strange that they did not make it part of the constructor

aslesarenko commented on src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala:
```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
+          .validateNoFailure(txDataBoxes, dataBoxesTry)
+          .payload[Long](validCostResult.value)
+          .validateTry(boxes, e => ModifierValidator.fatal("Missed data boxes", e)) { case (_, (dataBoxes, toSpend)) =>
+            tx.validateStateful(toSpend, dataBoxes, currentStateContext, validCostResult.value)(verifier).result
```
The currentStateContext is not update after each tx, is it ok?

pragmaxim replied:
it is, just a few lines above `costResult = tx.validateStateless()`
